frame/staking-rewards/src/lib.rs

```diff
@@ -599,8 +614,121 @@ pub mod pallet {
 					.try_insert(*asset_id, inflation)
 					.map_err(|_| Error::<T>::ReductionConfigProblem)?;
 			}
+
 			Ok((rewards_btree_map, reductions))
 		}
+
+		//TODO: tests
+		pub(crate) fn reward_acumulation_hook_reward_update_calculation(
+			pool_id: T::RewardPoolId,
+			asset_id: T::AssetId,
+			reward: Reward<T::AssetId, T::Balance>,
+			now: u64,
+		) -> (T::AssetId, Reward<T::AssetId, T::Balance>) {
+			let elapsed_time = match now.safe_sub(&reward.last_updated_timestamp) {
+				Ok(elapsed_time) => elapsed_time,
+				Err(_) => {
+					Self::deposit_event(Event::<T>::RewardAccumulationError { pool_id, asset_id });
+					return (asset_id, reward)
+				},
+			};
+
+			// SAFETY(benluelo): RewardRate::period is non-zero.
+			let periods_surpassed = elapsed_time.div(&reward.reward_rate.period);
+
+			if periods_surpassed.is_zero() {
+				(asset_id, reward)
+			} else {
+				// NOTE(benluelo): used this instead of `safe_mul` since
+				// `T::Balance` can't be multiplied by a `u64` and the definition of
+				// `CheckedMul` constrains the `Rhs` type to `Self`, so `T::balance`
+				// can't `.safe_mul(u64)`
+				//
+				// this also allows us to accumulate up to the maximum amount before
+				// erroring, short-circuiting on any error found with the last known
+				// "good" value.
+				let new_total_rewards = iter::repeat(reward.reward_rate.amount)
+					// REVIEW(benluelo): this cast is *probably* safe, but
+					// should be verified somehow
+					.take(periods_surpassed as usize)
+					.try_fold(reward.total_rewards, |total_amount, amount| {
+						match total_amount
+							.safe_add(&amount)
+							.map(|new_rewards| (new_rewards, new_rewards <= reward.max_rewards))
+						{
+							Ok((new_rewards, true)) => Ok(new_rewards),
+							// max rewards hit
+							Ok((new_rewards, false)) => Ok(new_rewards),
+							// overflow
+							Err(_) => Err(total_amount),
+						}
+					});
+
+				let new_total_rewards = match new_total_rewards {
+					Ok(amount) => amount,
+					Err(amount) => {
+						Self::deposit_event(Event::<T>::RewardAccumulationError {
+							pool_id,
+							asset_id,
+						});
+						amount
+					},
+				};
+
+				(
+					asset_id,
+					Reward {
+						total_rewards: new_total_rewards,
+						last_updated_timestamp: now,
+						..reward
+					},
+				)
+			}
+		}
+
+		//TODO: tests
+		pub(crate) fn acumulate_rewards_hook() -> Weight {
+			let now = T::UnixTime::now().as_secs();
+
+			RewardPools::<T>::iter()
+				.into_iter()
+				.map(|(pool_id, reward_pool)| {
+					let updated_rewards = reward_pool
+						.rewards
+						.into_iter()
+						.map(|(asset_id, reward)| {
+							Self::reward_acumulation_hook_reward_update_calculation(
+								pool_id, asset_id, reward, now,
+							)
+						})
+						.collect::<BTreeMap<_, _>>()
+						.try_into()
```

Comment:
`.into_iter() -> ... -> .try_into()` can be rewritten using [try_mutate()](https://paritytech.github.io/substrate/master/frame_support/struct.BoundedBTreeMap.html#method.try_mutate)

Reply:
I would still need to iterate over the map in the closure provided to try_mutate, this felt cleaner

Reply:
also see https://github.com/paritytech/substrate/pull/11869